api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java

```diff
@@ -0,0 +1,153 @@
+  private final Clock clock;
+
+  @Autowired
+  OfflineClusterController(
+      @Qualifier(NotebooksConfig.SERVICE_CLUSTER_API) Provider<ClusterApi> clusterApiProvider,
+      Provider<WorkbenchConfig> configProvider,
+      Clock clock) {
+    this.clock = clock;
+    this.configProvider = configProvider;
+    this.clusterApiProvider = clusterApiProvider;
+  }
+
+  /**
+   * checkClusters deletes older clusters in order to force an upgrade on the
+   * next researcher login. This method is meant to be restricted to invocation
+   * by App Engine cron.
+   *
+   * The cluster deletion policy here aims to strike a balance between enforcing
+   * upgrades, cost savings, and minimizing user disruption. To this point, our
+   * goal is to only upgrade idle clusters if possible, as doing so gives us an
+   * assurance that a researcher is not actively using it. We delete clusters in
+   * the following cases:
+   *
+   * 1. It exceeds the max cluster age. Per environment, but O(weeks).
+   * 2. It is idle and exceeds the max idle cluster age. Per environment,
+   *    smaller than (1).
+   *
+   * As an App Engine cron endpoint, the runtime of this method may not exceed
+   * 10 minutes.
+   */
+  @Override
+  public ResponseEntity<CheckClustersResponse> checkClusters() {
+    Instant now = clock.instant();
+    WorkbenchConfig config = configProvider.get();
+    Duration maxAge = Duration.ofDays(config.firecloud.clusterMaxAgeDays);
+    Duration idleMaxAge = Duration.ofDays(config.firecloud.clusterIdleMaxAgeDays);
+
+    ClusterApi clusterApi = clusterApiProvider.get();
+    List<Cluster> clusters;
+    try {
+      clusters = clusterApi.listClusters(null, false);
+    } catch (ApiException e) {
+      throw ExceptionUtils.convertNotebookException(e);
+    }
+
+    int errors = 0;
+    int activeDeletes = 0;
+    int unusedDeletes = 0;
+    for (Cluster c : clusters) {
+      String clusterId = c.getGoogleProject() + "/" + c.getClusterName();
+      try {
+        // Refetch the cluster to ensure freshness as this iteration may take
+        // some time.
+        c = clusterApi.getCluster(c.getGoogleProject(), c.getClusterName());
+      } catch (ApiException e) {
+        log.log(Level.WARNING, String.format("failed to refetch cluster '%s'", clusterId), e);
+        errors++;
+        continue;
+      }
+      if (ClusterStatus.UNKNOWN.equals(c.getStatus()) || c.getStatus() == null) {
+        log.warning(String.format("unknown cluster status for cluster '%s'", clusterId));
+        continue;
+      }
+      if (!ClusterStatus.RUNNING.equals(c.getStatus()) &&
+          !ClusterStatus.STOPPED.equals(c.getStatus())) {
+        // For now, we only handle running or stopped (suspended) clusters.
+        continue;
+      }
+
+      Instant created = Instant.parse(c.getCreatedDate());
+      Duration age = Duration.between(created, now);
+      Instant lastUsed = Instant.parse(c.getDateAccessed());
```

Comment:
When exactly does the dateAccessed field get updated? If I open a notebook and keep the notebook open, does it periodically update this value even if nothing is running? Or is it only if I"m running cells? What if the kernel is doing a long-running operation and I'm not doing anything in the UI?

Reply:
My understanding is that it is purely based on the most recent web request coming through the proxy, i.e. any traffic to the Jupyter API from the notebooks UI. It's unclear to me whether a tab open in the background would trigger this, or just typing (but not executing), or if only executing cells would trigger it. I could also imagine this varying between browsers. Asked on the Leo doc for clarification.

Ultimately I believe this value is the same one that Leo is using to trigger pause, and eventually we'll just switch this code to look for the STOPPED status instead of this timestamp. In other words, if this date field doesn't match the behavior we want, then autopause may not match the behavior we want.